File: src/main/java/org/fountainmc/api/entity/data/PlayerData.java
package org.fountainmc.api.entity.data;

import java.util.UUID;
import javax.annotation.Nonnegative;

import org.fountainmc.api.Fountain;
import org.fountainmc.api.GameMode;
import org.fountainmc.api.NonnullByDefault;

import static com.google.common.base.Preconditions.*;

/**
 * The data of a player.
 */
@NonnullByDefault
public interface PlayerData extends LivingEntityData {

    /**
     * Get the name of the Player.
     *
     * @return name of the Player
     */
    String getName();

    /**
     * Get the Unique Identifier of the Player.
     *
     * @return UUID of the Player
     */
    UUID getUniqueId();

    /**
     * Get the game-mode the player is (or will be) in
     *
     * @return the game-mode
     */
    GameMode getGameMode();

    void setGameMode(GameMode gameMode);

    boolean isFlying();

    void setFlying(boolean flying);

    boolean isFlyingAllowed();

    void setFlyingAllowed(boolean canFly);

    @Nonnegative
    float getPercentageToNextExperienceLevel();

    void setPercentageToNextExperienceLevel(@Nonnegative float percentage);

    @Nonnegative
    int getExperienceLevel();

    void setExperienceLevel(@Nonnegative int experienceLevel);

    default ExperienceData getExperienceData() {
        return ExperienceData.create(getExperienceLevel(), getPercentageToNextExperienceLevel());
    }

    default void setExperienceData(ExperienceData experienceData) {
        checkNotNull(experienceData, "Null experience data");
        setExperienceLevel(experienceData.getExperienceLevel());
        setPercentageToNextExperienceLevel(experienceData.getPercentageUntilNextExperienceLevel());
    }

    /**
     * Calculate and return the player's total experience
     * <p>The experience in each level is given in </p>
     *
     * @return the player's total experience
     */
    @Nonnegative
    default long getTotalExperience() {
        return getExperienceData().getTotalExperience();
    }

    default void setTotalExperience(long total) {
        setExperienceData(ExperienceData.fromTotalExperience(total));
    }

    default void giveExp(@Nonnegative long amount) {
        checkArgument(amount >= 0, "Can't give negative exp %s! Use takeExp(long) if you want that!", amount);
        setTotalExperience(getTotalExperience() + amount);
    }

    default void takeExp(@Nonnegative long amount) {
        checkArgument(amount >= 0, "Can't give negative exp %s! Use takeExp(long) if you want that!", amount);
        long exp = getTotalExperience();
        checkState(exp >= amount, "Can't take away %s exp because the player only has %s exp.", amount, exp);
        setTotalExperience(exp - amount);
    }

    @Override
    default void copyDataFrom(EntityData data) {
        LivingEntityData.super.copyDataFrom(data);
        if (data instanceof PlayerData) {
            this.setGameMode(((PlayerData) data).getGameMode());
            this.setFlying(((PlayerData) data).isFlying());
            this.setFlyingAllowed(((PlayerData) data).isFlyingAllowed());
            this.setPercentageToNextExperienceLevel(((PlayerData) data).getPercentageToNextExperienceLevel());
            this.setExperienceLevel(((PlayerData) data).getExperienceLevel());
        }
    }

    final class ExperienceData {
        public static final ExperienceData ZERO = ExperienceData.create(0, 0);
        private final float percentageUntilNextExperienceLevel;
        private final int experienceLevel;

        private ExperienceData(int experienceLevel, float percentageUntilNextExperienceLevel) {
            this.percentageUntilNextExperienceLevel = percentageUntilNextExperienceLevel;
            checkArgument(percentageUntilNextExperienceLevel < 1, "Percentage until next experience level %s is greater than or equal to 1", percentageUntilNextExperienceLevel);
            checkArgument(percentageUntilNextExperienceLevel >= 0, "Percentage until next experience level %s is negative", percentageUntilNextExperienceLevel);
            checkArgument(experienceLevel >= 0, "Experience level %s is negative", experienceLevel);
            this.experienceLevel = experienceLevel;
        }

        public float getPercentageUntilNextExperienceLevel() {
            return percentageUntilNextExperienceLevel;
        }

        public int getExperienceLevel() {
            return experienceLevel;
        }

        public long getTotalExperience() {
            return Fountain.getServer().calculateTotalExperience(this);
        }

        public static ExperienceData create(int experienceLevel, float percentageUntilNextExperienceLevel) {
            return new ExperienceData(experienceLevel, percentageUntilNextExperienceLevel);
        }

        public static ExperienceData fromTotalExperience(long totalExperience) {
            checkArgument(totalExperience >= 0, "Negative total experience %s", totalExperience);
            return Fountain.getServer().calculateExperienceData(totalExperience);
        }
    }
}

File: src/main/kotlin/org/fountainmc/api/entity/data/PlayerData.kt
package org.fountainmc.api.entity.data

import org.fountainmc.api.Fountain
import org.fountainmc.api.GameMode
import java.util.*

/**
 * The data of a player.
 */

interface PlayerData : LivingEntityData {

    /**
     * Get the name of the Player.

     * @return name of the Player
     */
    val name: String

    /**
     * Get the Unique Identifier of the Player.

     * @return UUID of the Player
     */
    val uniqueId: UUID

    /**
     * Get the game-mode the player is (or will be) in

     * @return the game-mode
     */
    var gameMode: GameMode

    var isFlying: Boolean

    var isFlyingAllowed: Boolean

    var percentageToNextExperienceLevel: Float

    var experienceLevel: Int

    var experienceData: ExperienceData
        get() = ExperienceData.create(experienceLevel, percentageToNextExperienceLevel)
        set(experienceData) {
            experienceData
            experienceLevel = experienceData.experienceLevel
            percentageToNextExperienceLevel = experienceData.percentageUntilNextExperienceLevel
        }

    /**
     * Calculate and return the player's total experience
     *
     * The experience in each level is given in

     * @return the player's total experience
     */
    var totalExperience: Long
        get() = experienceData.totalExperience
        set(total) {
            experienceData = ExperienceData.fromTotalExperience(total)
        }

    fun giveExp(amount: Long) {
        require(amount >= 0) { "Can't give negative exp $amount! Use takeExp if you want that!" }
        totalExperience += amount
    }

    fun takeExp(amount: Long) {
        require(amount >= 0) { "Can't take negative exp $amount! Use giveExp if you want that!" }
        val exp = totalExperience
        require(totalExperience >= amount) { "Can't take away $amount exp because the player only has $exp exp!" }
        totalExperience = exp - amount
    }

    override fun copyDataFrom(data: EntityData) {
        super.copyDataFrom(data)
        if (data is PlayerData) {
            this.gameMode = data.gameMode
            this.isFlying = data.isFlying
            this.isFlyingAllowed = data.isFlyingAllowed
            this.percentageToNextExperienceLevel = data.percentageToNextExperienceLevel
            this.experienceLevel = data.experienceLevel
        }
    }

    class ExperienceData private constructor(val experienceLevel: Int, val percentageUntilNextExperienceLevel: Float) {

        init {
            require(percentageUntilNextExperienceLevel < 1) {
                "Percentage until next experience level must be less than one: $percentageUntilNextExperienceLevel"
            }
            require(percentageUntilNextExperienceLevel >= 0) {
                "Percentage until next experience level can't be negative: $percentageUntilNextExperienceLevel"
            }
            require(experienceLevel >= 0) { "Experience level can't be negative: $experienceLevel" }
        }

        val totalExperience: Long
            get() = Fountain.calculateTotalExperience(this)

        companion object {
            val ZERO = ExperienceData.create(0, 0f)

            fun create(experienceLevel: Int, percentageUntilNextExperienceLevel: Float): ExperienceData {
                return ExperienceData(experienceLevel, percentageUntilNextExperienceLevel)
            }

            fun fromTotalExperience(totalExperience: Long): ExperienceData {
                require(totalExperience >= 0) { "Negative total experience $totalExperience" }
                return Fountain.calculateExperienceData(totalExperience)
            }
        }
    }
}
